File: app/domain/performance_metrics.py
import pandas as pd
import math
import logging
from app.core.logger import Logger


class Performance_Metrics:
    def __init__(self):
        self.logger = Logger(name="Performance Metrics", level=logging.DEBUG).logger
    
    def calculate_race_effort(self, distance: float, elevation: float) -> float:
        """
        
        """
        self.logger.debug(f"Calculating race effort for distance: {distance} and elevation: {elevation}")
        race_effort = distance + (elevation / 100)
        return race_effort
    
    def calculate_pace(self, distance: float, total_minutes: float) -> float:
        """

        """
        self.logger.debug(f"Calculating pace for distance: {distance} with total minutes: {total_minutes}")
        pace = total_minutes / distance
        return pace

    def calculate_vertical_rate(self, distance: float, elevation: float) -> float:
        """
        
        """
        self.logger.debug(f"Calculating vertical rate for distance {distance} and elevation: {elevation}")
        vertical_rate = elevation / distance
        return vertical_rate

    def calculate_estimated_pace_on_flat_equivalent(self, total_time: float, race_effort: float) -> float:
        """
        Estimated pace in minutes
        """
        self.logger.debug(f"Calculating estimated pace on flat equivalent for race effort {race_effort} with total time {total_time}")
        estimated_pace = total_time / race_effort
        return estimated_pace

    def calculate_race_difficulty_score(self, distance: float, elevation: float, alpha: float = 0.7, c: float = 80, k: float = 0.05) -> float:
        """
        score between 1 - 100
        """
        race_effort = self.calculate_race_effort(distance=distance, elevation=elevation)
        vertical_rate = self.calculate_vertical_rate(distance=distance, elevation=elevation)
        raw_score = race_effort + alpha * vertical_rate
        return 100 / (1 + math.exp(-k * (raw_score - c)))
    
    def calculate_race_category(self, distance: float, elevation: float) -> str:
        """
        
        """
        race_effort = self.calculate_race_effort(distance=distance, elevation=elevation)
        if race_effort < 20:
            return f"{race_effort}K"
        elif race_effort < 50:
            return "20K"
        elif race_effort < 100:
            return "50K"
        elif race_effort < 170:
            return "100K"
        return "100M"

    def generate_category_label(self, distance: float, elevation: float) -> float:
        """
        Flat
        Rolling
        Hilly
        Mountainous
        Extreme Mountain
        """
        vertical_rate = self.calculate_vertical_rate(distance=distance, elevation=elevation)
        if vertical_rate < 15:
            return "Flat"
        elif vertical_rate >= 15 and vertical_rate < 35:
            return "Rolling"
        elif vertical_rate >= 35 and vertical_rate < 55:
            return "Hilly"
        elif vertical_rate >= 55 and vertical_rate < 80:
            return "Mountain"
        elif vertical_rate >= 80:
            return "Extreme Mountain"

    def calculate_vertical_per_hour(self, total_hours: float, elevation_gain: float) -> float:
        """
        elevation_gain / total_hours
        """
        self.logger.debug(f"Calculating total vertical meters per hour with {elevation_gain} total meters in {total_hours} hours")
        vertical_per_hour = elevation_gain / total_hours
        return vertical_per_hour
    
    def calculate_all(self, distance: float, elevation: float, total_minutes: float, alpha: float = 0.7, c: float = 80, k: float = 0.05) -> dict:
        """
        
        """
        race_effort = self.calculate_race_effort(distance=distance, elevation=elevation)
        pace = self.calculate_pace(distance=distance, total_minutes=total_minutes)
        vertical_rate = self.calculate_vertical_rate(distance=distance, elevation=elevation)
        pace_on_flat_equivalent = self.calculate_estimated_pace_on_flat_equivalent(total_time=total_minutes, race_effort=race_effort)
        race_difficulty_score = self.calculate_race_difficulty_score(distance=distance, elevation=elevation, alpha=alpha, c=c, k=k)
        race_category = self.calculate_race_category(distance=distance, elevation=elevation)
        category_label = self.generate_category_label(distance=distance, elevation=elevation)
        return {
            "Race_effort": race_effort,
            "Pace": pace,
            "Vertical_rate": vertical_rate,
            "Pace_on_flat_equivalent": pace_on_flat_equivalent,
            "Race_difficulty_score": race_difficulty_score,
            "Race_category": race_category,
            "Category_label": category_label
        }

File: frontend/js/performanceMetrics.js


const performanceMetricsUrl = "http://127.0.0.1:8000/api/v1/performance_metrics/get_all";


const effortBasedRaceMatchingUrl = "http://127.0.0.1:8000/api/v1/performance_projection/effort_based_race_matching";
let effortBasedRaceMatchingcurrentPage = 1;
let effortBasedRaceMatchingPageSize = 10;
let effortBasedRaceMatchingHasNextPage = false;


const performanceBasedRaceMatchingUrl = "http://127.0.0.1:8000/api/v1/performance_projection/race_placement_projection";
let performanceBasedRaceMatchingCurrentPage = 1;
let performanceBasedRaceMatchingPageSize = 10;
let performanceBasedRaceMatchingHasNextPage = false;

const effortBasedRaceMatchingTable = document.querySelector("#effort-based-matches-table");

const form = document.querySelector("#performance-metrics-form");
const distanceInput = document.querySelector("#distance");
const elevationInput = document.querySelector("#elevation");
const totalTimeInput = document.querySelector("#total-time");

async function getAllPerformanceMetrics() {
    const distance = Number(distanceInput.value);
    const elevation = Number(elevationInput.value);
    const totalTime = Number(totalTimeInput.value);
    try {
        const response = await fetch(performanceMetricsUrl, {
            method: "POST",
            headers: {
                "Content-Type": "application/json"
            },
            body: JSON.stringify({
                distance: distance,
                elevation: elevation,
                total_minutes: totalTime
            })
        });
        if (!response.ok) {
            throw new Error("Request failed");
        }
        const data = await response.json();
        return data
    } catch (error) {
        console.error(error);
    }
}

async function getEffortBasedRaceMatching(page = 1) {
    const distance = Number(distanceInput.value);
    const elevation = Number(elevationInput.value);

    const limit = effortBasedRaceMatchingPageSize;
    const offset = (page - 1) * limit;
    try {
        const response = await fetch(effortBasedRaceMatchingUrl, {
            method: "POST",
            headers: {
                "Content-Type": "application/json"
            },
            body: JSON.stringify({
                distance: distance,
                elevation: elevation,
                limit: limit,
                offset: offset
            })
        });
        if (!response.ok) {
            throw new Error("Request failed");
        }
        const data = await response.json();
        return data
    } catch (error) {
        console.error(error);
    }
}


function renderRaceTable(races, tableBody) {
    const offset = (effortBasedRaceMatchingcurrentPage - 1) * effortBasedRaceMatchingPageSize;

    tableBody.innerHTML = races.map((race, index) => `
        <tr>
            <td>#${offset + index + 1}</td>
            <td>${race["Race_Title"]}</td>
            <td>${race["Race_Category"]}</td>
            <td>${race["Distance"]} km</td>
            <td>${race["Elevation_Gain"]} m+</td>
        </tr>
    `).join("");
}

async function loadRacePage(page = 1) {
    try {
        const data = await getEffortBasedRaceMatching(page);

        const races = data.result;

        effortBasedRaceMatchingcurrentPage = page;
        effortBasedRaceMatchingHasNextPage =
            races.length === effortBasedRaceMatchingPageSize;

        renderRaceTable(races, effortBasedRaceMatchingTable);

        document.querySelector("#effort-based-matching-prev-page-btn").disabled =
            effortBasedRaceMatchingcurrentPage <= 1;

        document.querySelector("#effort-based-matching-next-btn").disabled =
            !effortBasedRaceMatchingHasNextPage;

    } catch (error) {
        console.error(error);
    }
}


function convertPaceToReadablePace(pace) {
    const minutes = Math.floor(pace);
    let seconds = Math.round((pace - minutes) * 60);

    if (seconds === 60) {
        seconds = 0;
        minutes += 1;
    }
    return `${minutes}:${String(seconds).padStart(2, "0")}`;
}

function setRaceCategoryTextColor(raceCategoryItem, raceCategory) {
    if (raceCategory == '20K') {
        return "var(--utmb-20k)";
    }
    else if (raceCategory == '50K') {
        return "var(--utmb-50k)"
    }
    else if (raceCategory == '100K') {
        return "var(--utmb-100k)"
    }
    else if (raceCategory == '100M') {
        return "var(--utmb-100m)"
    }
    else {
        return "white";
    }
}

function setPercentageBar(raceDifficultyScoreItem, raceDifficultyScore, raceDifficultyScoreFillItem) {
    const safeValue = Math.max(0, Math.min(100, raceDifficultyScore));
    const finalValue = safeValue.toFixed(2);
    raceDifficultyScoreItem.textContent = finalValue;
    raceDifficultyScoreFillItem.style.width = `${finalValue}%`;
}

form.addEventListener("submit", async function (event) {
    event.preventDefault();

    const performanceMetrics = await getAllPerformanceMetrics();

    const raceEffort = performanceMetrics.results["Race_effort"];
    const pace = convertPaceToReadablePace(performanceMetrics.results["Pace"]);
    const verticalRate = performanceMetrics.results["Vertical_rate"].toFixed(2);
    const paceOnFlatEquivalent = convertPaceToReadablePace(performanceMetrics.results["Pace_on_flat_equivalent"]);
    const raceCategory = performanceMetrics.results["Race_category"];
    const raceDifficultyScore = performanceMetrics.results["Race_difficulty_score"];
    const categoryLabel = performanceMetrics.results["Category_label"];

    const raceEffortItem = document.querySelector("#race-effort");
    raceEffortItem.textContent = `${raceEffort} km`;
    
    const paceItem = document.querySelector("#pace");
    paceItem.textContent = `${pace} min/km`;

    const verticalRateItem = document.querySelector("#vertical-rate");
    verticalRateItem.textContent = `${verticalRate} m+/km`;

    const paceOnFlatEquivalentItem = document.querySelector("#pace-on-flat-equivalent");
    paceOnFlatEquivalentItem.textContent = `${paceOnFlatEquivalent} min/km`;

    const raceCategoryItem = document.querySelector("#race-category");
    raceCategoryItem.textContent = raceCategory;
    raceCategoryItem.style.color = setRaceCategoryTextColor(raceCategoryItem, raceCategory);

    const raceDifficultyScoreItem = document.querySelector("#race-difficulty-score");
    const raceDifficultyScoreFillItem = document.querySelector("#race-difficulty-score-fill");
    setPercentageBar(raceDifficultyScoreItem, raceDifficultyScore, raceDifficultyScoreFillItem);

    const categoryLabelItem = document.querySelector("#category-label");
    categoryLabelItem.textContent = categoryLabel;


    const effortBasedRaceMatching = await getEffortBasedRaceMatching(1);
    renderRaceTable(effortBasedRaceMatching.result, effortBasedRaceMatchingTable);

    await loadRacePage(1);
});

document.querySelector("#effort-based-matching-prev-page-btn")
    .addEventListener("click", () => {
        if (effortBasedRaceMatchingcurrentPage > 1) {
            loadRacePage(effortBasedRaceMatchingcurrentPage - 1);
        }
    });

document.querySelector("#effort-based-matching-next-btn")
    .addEventListener("click", () => {
        if (effortBasedRaceMatchingHasNextPage) {
            loadRacePage(effortBasedRaceMatchingcurrentPage + 1);
        }
    });